### PR TITLE
feat(cli): Use well-known endpoint to resolve API

### LINKED ISSUE #6487: [CLI] `immich login-key` is a little fussy when it comes to the `instanceUrl` provided
### The bug

One issue for a couple of problems I've noticed with the `login-key` command of the CLI.

1. Providing a link to the root immich instance (eg. `https://demo.immich.app/`) doesn't fail:

```console
root@5bc318f4d2ab:/work# immich login-key https://demo.immich.app/                        
Executing API key auth flow...
Logged in as undefined
Wrote auth info to /root/.config/immich/auth.yml
```

Despite not providing an API key it believe it's authenticated correctly as `undefined`. This is probably because it's getting a 200 OK from the login page.

Perhaps a suggestion to append `/api` could be useful if we get a response of `Content-Type: text/html`.

2. Providing an instanceUrl with a trailing slash causes things to break

```console
root@5bc318f4d2ab:/work# immich login-key https://demo.immich.app/api/ 123                                      
Executing API key auth flow...
/usr/local/lib/node_modules/@immich/cli/dist/src/services/session.service.js:44
            throw new login_error_1.LoginError(`Failed to connect to server ${instanceUrl}: ${error.message}`);
                  ^

LoginError: Failed to connect to server https://demo.immich.app/api/: Request failed with status code 404
    at /usr/local/lib/node_modules/@immich/cli/dist/src/services/session.service.js:44:19
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async SessionService.keyLogin (/usr/local/lib/node_modules/@immich/cli/dist/src/services/session.service.js:43:36)
    at async LoginKey.run (/usr/local/lib/node_modules/@immich/cli/dist/src/commands/login/key.js:7:9)
    at async Command.<anonymous> (/usr/local/lib/node_modules/@immich/cli/dist/src/index.js:57:5)

Node.js v21.6.0
```

Perhaps we could make a nicer error here (as described in #5357) and allow for a trailing slash (or not) to be provided and handle it gracefully?

### The OS that Immich Server is running on

Debian 11

### Version of Immich Server

1.92.0

### Version of Immich Mobile App

N/A

### Platform with the issue

- [ ] Server
- [ ] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
Not relevant
```

### Your .env content

```Shell
Not relevant
```

### Reproduction steps

```bash
Described above
```

### Additional information

_No response_

### LINKED ISSUE #6731: Unable to use immich CLI in docker
### The bug

I followed the official docker compose install guide and everything went well. I now want to use the cli to bulk import images into immich. I am not sure if I am doing something wrong, or if this is a bug.

Open a shell into the docker container and run:

```
immich upload -r -n .
No auth file exist. Please login first

immich login-key 127.0.0.1:3001 Ix.....JhA # 3001 is the internal port used by immich


Executing API key auth flow...
/usr/src/app/node_modules/@immich/cli/dist/src/services/session.service.js:44
            throw new login_error_1.LoginError(`Failed to connect to server ${instanceUrl}: ${error.message}`);
                  ^

LoginError: Failed to connect to server 127.0.0.1:3001: connect ECONNREFUSED 127.0.0.1:80
    at /usr/src/app/node_modules/@immich/cli/dist/src/services/session.service.js:44:19
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async SessionService.keyLogin (/usr/src/app/node_modules/@immich/cli/dist/src/services/session.service.js:43:36)
    at async LoginKey.run (/usr/src/app/node_modules/@immich/cli/dist/src/commands/login/key.js:7:9)
    at async Command.<anonymous> (/usr/src/app/node_modules/@immich/cli/dist/src/index.js:57:5)


immich login-key http://127.0.0.1:3001 Ixb.....JhA
Executing API key auth flow...
Logged in as undefined
Wrote auth info to /root/.config/immich/auth.yml

# That kind of worked, but my name is admin, and I have logged in multiple times with the web UI. Not sure why I am called undefined here

# I then try to dry run the upload command on everything in the current dir. I am currently in a bind mounted dir full of pictures

immich upload -r -n .
/usr/src/app/node_modules/@immich/cli/dist/src/services/session.service.js:68
            throw new Error(`Could not parse response. Is Immich listening on ${this.api.apiConfiguration.instanceUrl}?`);
                  ^

Error: Could not parse response. Is Immich listening on http://127.0.0.1:3001?
    at SessionService.ping (/usr/src/app/node_modules/@immich/cli/dist/src/services/session.service.js:68:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async SessionService.connect (/usr/src/app/node_modules/@immich/cli/dist/src/services/session.service.js:38:9)
    at async Upload.connect (/usr/src/app/node_modules/@immich/cli/dist/src/cli/base-command.js:16:30)
    at async Upload.run (/usr/src/app/node_modules/@immich/cli/dist/src/commands/upload.js:16:9)
    at async Command.<anonymous> (/usr/src/app/node_modules/@immich/cli/dist/src/index.js:43:5)

Node.js v20.11.0
```

### The OS that Immich Server is running on

Official immich docker container (Debian GNU/Linux 12)

### Version of Immich Server

v1.93.3

### Version of Immich Mobile App

Not used

### Platform with the issue

- [X] Server
- [ ] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
version: "3.8"

name: immich

services:
  immich-server:
    container_name: immich_server
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION:-release}
    command: [ "start.sh", "immich" ]
    volumes:
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - /etc/localtime:/etc/localtime:ro
      - /media/immich_temp_pics:/immich_temp_pics:ro
    env_file:
      - .env
    ports:
      - 2283:3001
    depends_on:
      - redis
      - database
    restart: unless-stopped

  immich-microservices:
    container_name: immich_microservices
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION:-release}
    # extends:
    #   file: hwaccel.yml
    #   service: hwaccel
    command: [ "start.sh", "microservices" ]
    volumes:
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - /etc/localtime:/etc/localtime:ro
      - /media/immich_temp_pics:/immich_temp_pics:ro
    env_file:
      - .env
    depends_on:
      - redis
      - database
    restart: unless-stopped

  immich-machine-learning:
    container_name: immich_machine_learning
    image: ghcr.io/immich-app/immich-machine-learning:${IMMICH_VERSION:-release}
    volumes:
      - model-cache:/cache
    env_file:
      - .env
    restart: unless-stopped

  redis:
    container_name: immich_redis
    image: redis:6.2-alpine@sha256:c5a607fb6e1bb15d32bbcf14db22787d19e428d59e31a5da67511b49bb0f1ccc
    restart: unless-stopped

  database:
    container_name: immich_postgres
    image: tensorchord/pgvecto-rs:pg14-v0.1.11@sha256:0335a1a22f8c5dd1b697f14f079934f5152eaaa216c09b61e293be285491f8ee
    env_file:
      - .env
    environment:
      POSTGRES_PASSWORD: ${DB_PASSWORD}
      POSTGRES_USER: ${DB_USERNAME}
      POSTGRES_DB: ${DB_DATABASE_NAME}
    volumes:
      - ./db_data:/var/lib/postgresql/data
    restart: unless-stopped

volumes:
  model-cache:
```

### Your .env content

```Shell
# You can find documentation for all the supported env variables at https://immich.app/docs/install/environment-variables

# The location where your uploaded files are stored
UPLOAD_LOCATION=/media/immich

# The Immich version to use. You can pin this to a specific version like "v1.71.0"
IMMICH_VERSION=release

# Connection secret for postgres. You should change it to a random password
DB_PASSWORD=secret

# The values below this line do not need to be changed
###################################################################################
DB_HOSTNAME=immich_postgres
DB_USERNAME=postgres
DB_DATABASE_NAME=immich

REDIS_HOSTNAME=immich_redis
```

### Reproduction steps

```bash
1. Run docker image according to instructions in docs
2. Shell into container
3. Run `immich login-key http://127.0.0.1:3001 <API key>`
4. Run `immich upload -r -n .` in a folder containing pictures
...
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by bo0tzz**
Fixes #6487

### CODE REVIEW

**Comment by bo0tzz** [comment]
Would fix #6487, #6731

**Comment by bo0tzz** [comment]
The fallback logic for constructing the URL is a bit more complex than I'd anticipated. For example, this code currently doesn't handle cases where .well-known returns successfully but builds an invalid endpoint.

**Comment by etnoy** [comment]
I'm not exactly sure what well-known does. Could you summarize the idea here?

**Comment by jrasm91** [comment]
The API is served at /api usually, but it you type in the root path / to the mobile app, there is a special endpoint (/.well-known/immich) that can check instead and it returns a response that indicates where the API is actually located. It is like a special discovery endpoint that can be used so the user isn't required to remember to add /api

File: cli/src/commands/login.command.ts [hunk 1]
```diff
@@ -2,6 +2,12 @@ import { BaseCommand } from './base-command';
 
 export class LoginCommand extends BaseCommand {
   public async run(instanceUrl: string, apiKey: string): Promise<void> {
+    if (!instanceUrl) {
```
Comment (jrasm91) [inline cli/src/commands/login.command.ts:5]: This should not be needed anymore.

**Review by jrasm91 — COMMENTED** [review]
You can probably more easily write an e2e test for this now btw.

**Review by jrasm91 — APPROVED** [review]
